### ISSUE TITLE
Implement logout functionality

### ISSUE THREAD

**Issue by natewalck** [macadmins/outset#10]
Would it be worth adding logout functionality ala https://github.com/aysiu/offset ?

**Comment by bartreardon** [comment]
Seams reasonable.

Detecting the state is...[something](https://github.com/aysiu/offset/blob/c1601b77d542513fb2bfa1052b0a05a998374ee1/offset#L213)

**Comment by natewalck** [comment]
There has to be a more macOS native way to do this. Hrmmm.

**Comment by natewalck** [comment]
Looking at Munki, it uses:

```
	<key>LimitLoadToSessionType</key>
	<array>
		<string>LoginWindow</string>
	</array>
```

Which is the obvious way to do it at the login window. I'd argue a reboot is also a logout, so we don't need to care about that (bur document it clearly)

System startup on the other hand is *not* a logout, so we'd need some method to figure that out. 

This lead me to finding this:

https://stackoverflow.com/a/63372640

So perhaps you could have a LaunchD that does nothing but run code to detect these events, then touch a file on disk that the `LoginWindow` LaunchAgent looks for (also similar to munki: https://github.com/munki/munki/blob/main/launchd/LaunchAgents/com.googlecode.munki.managedsoftwareupdate-loginwindow.plist#L9)

This of course circles right back around to what is the use case for this? Is it worth the time and effort? (I could see computer labs perhaps wanting something like this).

**Comment by bartreardon** [comment]
it was worth it enough for someone to write offset. The computer lab scenario of "run scripts that perform cleanup tasks" is one I can see people using. For that I don't think it's required to detect a logout event per-se, you just have a launchd that loads in the login window context.

Properly documented the workflow would be left to the admin, e,g,

loginwindow tasks always run at the login window. if you want an even to take place _after_ a logout then use something like this:
 - create your login window script to require presence of a trigger file
 - create a login-every script to create the trigger file
 - when the user logs out and you are back at the login window, your task should trigger, and then clean up after itself
 - optionally add a `boot` script that performs initial cleanup before the loginwindow is presented (e.g. your script trigger is still present)

I don't think it needs to be complicated. At a reboot event, the login window isn't displayed so these events won't trigger in that scenario. 

In short I think we don't implement "logout" functionality. We implement "login-window" and the logic for a "logout" scenario can be an exercise for the admin - maybe document some basic scenario as an example but not build that into outset.

WDYT?

**Comment by bartreardon** [comment]
this is also a core justification for Ouset IMO - creation of launchd logic for admin tasks that can be managed under a single banner and save admins from creating their own seperate ones and individually managing the login items settings.

**Comment by natewalck** [comment]
I like this method better. Less complicated and allows for as much flexibility as the admin might desire. 

> it was worth it enough for someone to write offset. The computer lab scenario of "run scripts that perform cleanup tasks" is one I can see people using. For that I don't think it's required to detect a logout event per-se, you just have a launchd that loads in the login window context.
> 
> Properly documented the workflow would be left to the admin, e,g,
> 
> loginwindow tasks always run at the login window. if you want an even to take place _after_ a logout then use something like this:
> 
> * create your login window script to require presence of a trigger file
> * create a login-every script to create the trigger file
> * when the user logs out and you are back at the login window, your task should trigger, and then clean up after itself
> * optionally add a `boot` script that performs initial cleanup before the loginwindow is presented (e.g. your script trigger is still present)
> 
> I don't think it needs to be complicated. At a reboot event, the login window isn't displayed so these events won't trigger in that scenario.
> 
> In short I think we don't implement "logout" functionality. We implement "login-window" and the logic for a "logout" scenario can be an exercise for the admin - maybe document some basic scenario as an example but not build that into outset.
> 
> WDYT?

**Comment by bartreardon** [comment]
Closing this one - while there's no direct "logout" function as such, this can be achieved with a `login-window` script and some flags set on user login to trigger the login window script to run at user logout (either as a logout, or reboot event)
